### ISSUE TITLE
Consider removing Mobile Safari empty onclick hack

### ISSUE THREAD

**Issue by gaearon** [facebook/react#12989]
See https://github.com/facebook/react/issues/238 and https://github.com/facebook/react/pull/1536 for historical context.

Is that still relevant? The code is here:

https://github.com/facebook/react/blob/52fbe7612e0527b8c86decac519c344626f6bd72/packages/react-dom/src/client/ReactDOMFiberComponent.js#L244-L245

Even if it's relevant, can we just feature test it, and not do this hack on other browsers? Seems like a waste of memory for event handlers (even though the function is the same every time).

**Comment by VideoCasino** [comment]
Needs investigation

**Comment by nhunzaker** [comment]
I'd like to limit this to Safari. Looks like we still need this for non-interactive elements. 

I did a few quick checks using:
https://codepen.io/nhunzaker/pen/yEMpJO?editors=1010

- This definitely is still an issue on iOS 11
- It looks like this isn't a problem in desktop Safari (7.1, 8, 11.1)

I wonder what is the best way to detect Safari (or generally when to apply this fix)

**Comment by gaearon** [comment]
Can we literally feature-test it? Like dispatching an event and seeing if it bubbles, or something?

**Comment by gaearon** [comment]
The problem with relying on useragent etc is that those can get messed up in webviews.

**Comment by nhunzaker** [comment]
Ah totally. I agree that we should feature detect this. I can check later today, but I wonder if it could be something like:

```
let noBubble = true

// where should document come from?

let parent = document.createElement('div')
let child = document.createElement('div')
let event = document.createEvent('Event')

event.initEvent('click', true, false)

parent.addEventListener('click', () => {
  noBubble = false
})

child.dispatchEvent(event);
```

**Comment by philipp-spiess** [comment]
The feature test using manual event dispatching does not seem to work ([here's what I used to test it](http://jsfiddle.net/96uy7rkh/14/)). I was thinking that this bail out maybe happens in Safari's hit-detection logic and we could feature detect using [`elementFromPoint()`](http://jsfiddle.net/9box1t8y/9/) but that also behaved correctly. 🤔

**Comment by hectorcoronado** [comment]
Hello! Would I be correct in surmising this may plausibly be a good first contribution? Haven't really dug into it sufficiently, but it looks doable. If so, and if one of you isn't already working on it, any chance you could assign to me?

**Comment by gaearon** [comment]
@hectorcoronado Feel free to dig in. This will need some research — it's great if you can do it!

**Comment by hectorcoronado** [comment]
I'll look into it in the coming days and update accordingly -- thanks!

**Comment by gaearon** [comment]
[Oh...](https://www.quirksmode.org/blog/archives/2014/02/mouse_event_bub.html)

>The target element, or **any of its ancestors** up to but not including the <body>, has an explicit event handler set for any of the mouse events. This event handler may be an empty function.

This explains why https://github.com/facebook/react/pull/11927 works. But it also means we might get away with only setting `onclick` on the root if the click delay / grey overlay issue is solved in version we support.

**Comment by philipp-spiess** [comment]
> But it also means we might get away with only setting `onclick` on the root if the click delay / grey overlay issue is solved in version we support.

To give some additional information: The "grey overlay" is a tap highlight that is shown when an element is pressed on iOS Safari. Safari will detect all interactive elements (e.g buttons) as tapable by default and adding an `onclick` handler will also make non-interactive elements (e.g div) tapable. 

It's important UX wise to respect the proper granularity for adding these listeners, otherwise the tap highlight will be shown above the whole container: 

![](https://dzwonsemrish7.cloudfront.net/items/0n3g0S0u2J0o0a1k2z1E/Screen%20Recording%202018-08-20%20at%2011.38%20AM.gif)

([Source code](https://codesandbox.io/s/xjjqzoyoqz))

--- 

Back to the original topic I just realized that this is tightly intertwined with the tap highlight color so maybe a feasible feature test would check if the browser supports that? Maybe we can check if the [`-webkit-tap-highlight-color`](https://developer.mozilla.org/de/docs/Web/CSS/-webkit-tap-highlight-color) CSS property is supported.

**Comment by philipp-spiess** [comment]
@hectorcoronado Do you still want to work on this issue? You can try to explore the CSS property based feature-detection if you want 😊 (See my comment above)

**Comment by hectorcoronado** [comment]
@philipp-spiess @gaearon yes! I will spend time on it after work tomorrow and will report back if it seems beyond my capabilities at present -- thank you!

**Comment by binhpv** [comment]
@gaearon 
I tested several iOS simulator and found out that from iOS 13 and higher, the click event bubbling issue is not an issue anymore. This test was done following test cases in https://www.quirksmode.org/blog/archives/2010/09/click_event_del.html

Maybe we could be more pragmatic and add the check only for iOS 12 and lower. This will of course cause some issues, but only for very small amount of user that are on iOS 12 or lower, and the user agent is modified and therefore we can't detect iOS anymore.

Currently our application is public sector application, and accessibility certificate is a must. But as in #13798, we are facing this "clickable" issue.

Another solution could be to provide a way to configure this behavior in the application code?

**Comment by gaearon** [comment]
Do we have a good way to detect iOS <= 13? I was just thinking about getting rid of this too a few days ago.

**Comment by binhpv** [comment]
@gaearon 
Using user agent detection like https://stackoverflow.com/questions/57599945/how-to-detect-ios-13-on-javascript, if we ignore those webview modified user agent :D

**Comment by gaearon** [comment]
I don’t think we can ignore webviews. The cost of the mistake is pretty high — things literally won’t be clickable.

**Comment by eps1lon** [comment]
> Back to the original topic I just realized that this is tightly intertwined with the tap highlight color so maybe a feasible feature test would check if the browser supports that? Maybe we can check if the -webkit-tap-highlight-color CSS property is supported.

FYI: Even desktop browsers implement `-webkit-tap-highlight-color` property and actually have values set for this property (`window.getComputedStyle(document.body).webkitTapHighlightColor`). So a naive `webkitTapHighlightColor in document.body.style` and even `window.getComputedStyle(document.body).webkitTapHighlightColor !== ''` would not tell you whether this color is applied when tapping.

Just setting `-webkit-tap-highlight-color: transparent` on React containers would also not work since that CSS property is unfortunately inherited i.e.  `#react-root button` would also have a transparent tap highlight color. Note that [MDN says it's not inherited](https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-tap-highlight-color) but that's not how Chrome 96 implements it: https://codesandbox.io/s/wizardly-vaughan-2cbh5?file=/src/App.js. Though we can prevent inheritance with `#react-root > * { -webkit-tap-highlight-color: initial; }`. 

So for clickable elements (elements with a click handler) we have to set an explicit `node.onclick` unless we find a robust way to implement browser heuristics. 

And even then it's questionable if we truly want to see the tap highlight color in case the click handler isn't even used to make that specific element interactive (e.g. tracking clicks, click-away detection)

With regards to calling click handlers:
The event delegation quirk only applies when only `document.body` has the click handler. If you have  a click handler on any other element then iOS will properly bubble click events up even when non-interactive elements are targeted. Tested on Safari iOS 9 (iPhone 6s): https://csb-sel0s-31qhjklet-eps1lon.vercel.app/ Note how only a click on the last fake button doesn't register. Every other click is registered because its (fake) react root has a click handler attached.

It doesn't look like we can safely remove this hack without affecting visual feedback on iOS which is unfortunate considering the empty click handler negatively affects NVDA + Firefox: https://github.com/nvaccess/nvda/issues/13108
